runtime/compiler/control/JitDump.cpp

```diff
@@ -153,29 +176,68 @@ jitDumpRecompileWithTracing(J9VMThread *vmThread, J9Method *ramMethod, TR::Compi
 
    // The request to use a trace log gets passed to the compilation via the optimization plan. The options object
    // created before the compile is issued will use the trace log we provide to initialize IL tracing.
-   TR_OptimizationPlan *plan = TR_OptimizationPlan::alloc(optLevel);
-   if (NULL == plan)
+   TR_OptimizationPlan *plan = NULL;
+#if defined(J9VM_OPT_JITSERVER)
+   // JITServer will create a new optimization plan on the client
+   if (compInfo->getPersistentInfo()->getRemoteCompilationMode() != JITServer::SERVER)
+#endif
       {
-      j9nls_printf(PORTLIB, J9NLS_INFO | J9NLS_STDERR, J9NLS_DMP_JIT_OPTIMIZATION_PLAN);
-      return;
-      }
+      plan = TR_OptimizationPlan::alloc(optLevel);
+      if (NULL == plan)
+         {
+         j9nls_printf(PORTLIB, J9NLS_INFO | J9NLS_STDERR, J9NLS_DMP_JIT_OPTIMIZATION_PLAN);
+         return;
+         }
 
-   plan->setInsertInstrumentation(isProfilingCompile);
-   plan->setLogCompilation(jitdumpFile);
+      plan->setInsertInstrumentation(isProfilingCompile);
+      plan->setLogCompilation(jitdumpFile);
+      }
 
    trfprintf(jitdumpFile, "<recompilation>\n");
 
    // Set the VM state of the crashed thread so the diagnostic thread can use consume it
    compInfo->setVMStateOfCrashedThread(vmThread->omrVMThread->vmState);
 
    J9::JitDumpMethodDetails details(ramMethod, optionsFromOriginalCompile, isAOTBody);
-   auto rc = compilationOK;
    auto queued = false;
-   compInfo->compileMethod(vmThread, details, oldStartPC, TR_no, &rc, &queued, plan);
+   auto rc = compilationOK;
+#if defined(J9VM_OPT_JITSERVER)
+   if (compInfo->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
+      {
+      // Inform the client that the compilation thread has crashed and send it
+      // a pointer to the JitDump file, so that it can request
+      // a JitDump recompilation from the server with an updated plan and method details
+      JITServer::ServerStream *stream = TR::compInfoPT->getMethodBeingCompiled()->_stream;
+      stream->write(JITServer::MessageType::compilationThreadCrashed, jitdumpFile);
+      stream->read<JITServer::Void>();
+      TR_MethodToBeCompiled *entry = NULL;
+         {
+         // Add an entry to the compilations queue using the current stream,
+         // and immediatelly set its method details as JitDump method details,
+         // so that only the diagnostic thread can compile it.
+         OMR::CriticalSection compilationMonitorLock(compInfo->getCompilationMonitor());
+         entry = compInfo->addOutOfProcessMethodToBeCompiled(stream);
```

Comment:
The diagnostic thread has already been resumed, might have found an empty queue and have gone to sleep on the compilation queue monitor waiting for work. After `addOutOfProcessMethodToBeCompiled` we must notify any such waiting compilation thread. See how we do it in `requeueOutOfProcessEntry`:

Reply:
I wonder if it's possible for another compilation thread to pick our diagnostic entry.
Another bad scenario:
1. Another thread is in `requeueOutOfProcessEntry` and has executed `addOutOfProcessMethodToBeCompiled`. At this point the thread has barely missed the `_disableFurtherCompilation` flag.
2. The crashing thread starts its procedure, resumes the diagnostic thread, which goes and picks the non-diagnostic compilation from the queue.

Reply:
This code makes sure that a non-diagnostic thread cannot pickup a diagnostic compilation:
https://github.com/eclipse/openj9/blob/eef164d0540b6576b45df84b4948b1f53b37a6a0/runtime/compiler/control/CompilationThread.cpp#L5231-L5272

Reply:
> The diagnostic thread has already been resumed, might have found an empty queue and have gone to sleep on the compilation queue monitor waiting for work. After `addOutOfProcessMethodToBeCompiled` we must notify any such waiting compilation thread. See how we do it in `requeueOutOfProcessEntry`:

Hm, I think now that I moved non-requeuing functionality out of `requeuOutOfProcessEntry` it can be called by the JitDump to queue an entry.